src/v/cloud_topics/level_zero/stm/ctp_stm.cc

```diff
@@ -112,8 +190,44 @@ ss::future<> ctp_stm::apply_raft_snapshot(const iobuf& buf) {
 }
 
 ss::future<iobuf> ctp_stm::take_raft_snapshot(model::offset snapshot_at) {
-    auto st = _state.get_state_at(snapshot_at);
-    co_return serde::to_iobuf(std::move(st));
+    vassert(
+      last_applied() >= snapshot_at,
+      "The snapshot is taken at offset {} but current insync offset is {}",
+      snapshot_at,
+      last_applied());
+    co_return serde::to_iobuf(_state);
+}
+
+ss::future<cluster_epoch_fence> ctp_stm::fence_epoch(cluster_epoch e) {
+    auto term = _raft->confirmed_term();
+    auto max_seen_epoch = _state.get_max_seen_epoch();
+    if (max_seen_epoch.has_value() && max_seen_epoch.value() == e) {
+        // Case 1. Same epoch, need to acquire read-lock.
+        auto unit = co_await _lock.hold_read_lock();
+        // Invariant: the max_seen_epoch is not nullopt because once
+        // set the max_seen_epoch is never resets.
+        if (_state.get_max_seen_epoch() == e) {
+            // The max_seen_epoch didn't advance after the scheduling point
+            co_return cluster_epoch_fence{std::move(unit), term};
+        }
+    } else {
+        // Case 2. New epoch, need to acquire write-lock.
+        auto unit = co_await _lock.hold_write_lock();
+        auto current_epoch = _state.get_max_seen_epoch();
+        if (!current_epoch.has_value() || current_epoch.value() <= e) {
+            _state.advance_max_seen_epoch(e);
+            unit.return_all();
+            // Invariant: we should be able to acquire units from the read lock
+            // here because we just released units from the write lock and there
+            // is no scheduling point in between we release and acquire the read
+            // lock.
```

Comment:
are semaphores in seastar not fair? This should still return nullopt if there are waiters

Reply:
https://github.com/scylladb/seastar/blob/master/include/seastar/core/semaphore.hh#L235

If there are waiters this will crash

Reply:
right